### PR TITLE
Support threads in Matrix (MSC3440)

### PR DESCRIPTION
**Pull request by jaller94**
Fixes https://github.com/matrix-org/matrix-appservice-slack/issues/634

Note: Slack doesn't seem to have a separate UI for replies. Everything is a thread.

Screenshot_2022-03-03_16-54-22

## Definition of Done
* [x] Slack threads are bridged to Matrix as `m.thread` relations – removing the reply relation we currently set.

### CODE REVIEW

**Comment by jaller94** [comment]
Known issues:

* [ ] Slack's feature "Also send to [main timeline]" isn't supported in Matrix. Messages will only show up in the thread.
* [ ] Sending images and other files does not work in threads.
  * [ ] Slack -> Matrix: The image arrives in the main timeline, not in the thread.
  * [ ] Matrix -> Slack: The message appears in the thread, but is empty.

**Comment by Half-Shot** [comment]
> Slack's feature "Also send to [main timeline]" isn't supported in Matrix. Messages will only show up in the thread.

Can we do a dumb thing where we just send the event twice to Matrix? Would this be too spammy? Alternatively should we comment on the proposal to raise the gap in the spec?

> Sending images and other files does not work in threads. 

I'd be reasonably okay with leaving this as an open issue to fix, if it's not an easy fix.

File: src/SlackGhost.ts [hunk 1]
```diff
@@ -312,6 +316,29 @@ export class SlackGhost {
...
+                "rel_type": "io.element.thread",
+                // If the reply event is part of a thread, continue the thread.
+                // Otherwise, attach a thread to the reply event.
+                "event_id": replyEvent.content["m.relates_to"]?.event_id ?? replyEvent.event_id,
+                "m.in_reply_to": {
+                    event_id: replyEvent.event_id,
+                },
+            },
+            "msgtype": "m.text", // for those who just want to send the reply as-is
+            "body": `${fallbackText}\n\n${this.prepareBody(text)}`,
```
Comment (Half-Shot) [inline src/SlackGhost.ts:335]: So I'm not sure if you still need a fallback or not, looking at what Element does:

image

File: changelog.d/634.feature [hunk 2]
```diff
@@ -0,0 +1 @@
+Support threads with m.thread relation (MSC3440)
```
Comment (Half-Shot) [inline changelog.d/634.feature:1]:
```suggestion
Support bridging Slack message threads with new m.thread relations (MSC3440)
```

Comment (Half-Shot) [reply changelog.d/634.feature:1]: (wanted to add a bit of colour)

**Review by Half-Shot — APPROVED** [review]
Code looks super clean and appears to work, nice one!

File: src/SlackGhost.ts (hunk 1 above)
Comment (jaller94) [reply src/SlackGhost.ts:335]: Element does not provide this fallback.
I guess if a client or bridge (e.g. IRC) wants to display the relation, it up to that cleint's or bridge's implementation to provide an adequate representation.

**Pull request merged**